### PR TITLE
Modified the quota cache refresh logic to keep element remain in the cache

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
I will file an another PR for api_manager to change the quota allocation mode from NORMAL to BEST_EFFORT

### CODE REVIEW

File: src/quota_aggregator_impl.cc [hunk 1]
```diff
@@ -55,6 +55,10 @@ QuotaAggregatorImpl::CacheElem::ReturnAllocateQuotaRequestAndClear(
     *(request.mutable_allocate_operation()) =
         operation_aggregator_->ToOperationProto();
     operation_aggregator_ = NULL;
+  } else {
+    // If requests ere not aggregated, use the stored initial request
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: s/ere/are/

File: src/quota_aggregator_impl.cc [hunk 2]
```diff
@@ -210,25 +214,19 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
...
+    // insert the element back to the cache
     cache_->Insert(elem->signature(), elem, 1);
 
-    AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
-        service_name_, service_config_id_));
+    if (elem->in_flight() == false) {
+      // refresh quota when the response is negative or aggregated.
+      if(elem->is_positive_response() == false || elem->is_aggregated()) {
+        elem->set_in_flight(true);
+        AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: This is not good.   Basically,  we are requesting 1 token in every second even there is not traffic.
I think  "ping server for token" in CLOSE mode should be triggered by traffic, not by timer.  It should be traffic + timer.      

When a request come,  check last_ping_time,   if > 1second,  send another ping.

File: src/quota_aggregator_impl.cc [hunk 3]
```diff
@@ -210,25 +228,16 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
...
-    // response.
-    elem->set_in_flight(true);
+  if (in_flush_all_ == false) {
+    // insert the element back to the cache
     cache_->Insert(elem->signature(), elem, 1);
 
-    AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
+    if (elem->in_flight() == false && elem->is_aggregated()) {
+      elem->set_in_flight(true);
+      AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: last_refresh_time should be set here.  at request sending time, not response time.

File: src/quota_aggregator_impl.cc [hunk 4]
```diff
@@ -137,6 +142,18 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
   // Aggregate tokens if the cached response is positive
   if (lookup.value()->is_positive_response()) {
     lookup.value()->Aggregate(request);
+  } else {
+    if (lookup.value()->in_flight() == false) {
+      int64_t elapsed = (SimpleCycleTimer::Now() - lookup.value()->last_refresh_time())
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: can we create a refresh_interval_in_cycle_ variable, just like

https://github.com/cloudendpoints/service-control-client-cxx/blob/master/src/check_aggregator_impl.cc#L65

Then compare the time stamp like

https://github.com/cloudendpoints/service-control-client-cxx/blob/master/src/check_aggregator_impl.cc#L166

File: src/quota_aggregator_impl.cc [hunk 5]
```diff
@@ -137,6 +142,18 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
   // Aggregate tokens if the cached response is positive
   if (lookup.value()->is_positive_response()) {
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: Here even for positive, we need to check last_refresh_time to make sure it is not too old

File: src/quota_aggregator_impl.cc [hunk 6]
```diff
@@ -137,6 +142,18 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
...
+  } else {
+    if (lookup.value()->in_flight() == false) {
+      int64_t elapsed = (SimpleCycleTimer::Now() - lookup.value()->last_refresh_time())
+          / 1000;
+      if (elapsed > options_.refresh_interval_ms) {
+        // update in_flight to avoid duplicated request
+        lookup.value()->set_in_flight(true);
+
+        // trigger a new quota refresh request
+        return Status(Code::NOT_FOUND, "");
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: We can use AddRemovedItem() to send refresh.  Not to using returning NOT_FOUND.

Just return the cached response.

File: src/quota_aggregator_impl.cc [hunk 7]
```diff
@@ -125,7 +129,8 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
     // requests will be aggregated to this temporary element until the
     // response for the actual request arrives.
     ::google::api::servicecontrol::v1::AllocateQuotaResponse temp_response;
-    CacheElem* cache_elem = new CacheElem(temp_response);
+    CacheElem* cache_elem = new CacheElem(request, temp_response,
+                                          SimpleCycleTimer::Now());
     cache_elem->set_signature(request_signature);
     cache_elem->set_in_flight(true);
     cache_->Insert(request_signature, cache_elem, 1);
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: Here even for the first time (new cache item,   we can just return temp_response,  and use AddRemovedItem() to send the refresh to sever. 

By doing this way, even the first request, we are not waiting for server respond

File: src/quota_aggregator_impl.cc (hunk 5 above)
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: So the logic in Quota() is

lookup_cache, if not_found,  {
   create a cache item
  AddRemovedItem() to send call to server
   return positive
}

Check last_refresh_time, if too old {
    AddRemoveItem() to send refresh time.
} else {
     if positive {
           aggregate
    }
}
return cached_response;

File: src/quota_aggregator_impl.cc (hunk 7 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: When the cached response is negative, exceeded timeout and quota was refilled; the client request will be rejected until the response arrives. Would this be ok?

File: src/quota_aggregator_impl.cc (hunk 5 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: Updated

File: src/quota_aggregator_impl.cc (hunk 4 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: Added ShouldRefresh method

File: src/quota_aggregator_impl.cc (hunk 6 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: Applied

File: src/quota_aggregator_impl.cc (hunk 3 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: Updated

File: src/quota_aggregator_impl.cc (hunk 5 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: When the cached response is negative, exceeded timeout and quota was refilled; the client request will be rejected until the response arrives. Would this be ok?

File: src/quota_aggregator_impl.cc (hunk 7 above)
Comment (mangchiandjjoe) [inline src/quota_aggregator_impl.cc]: Updated

File: src/quota_aggregator_impl.cc [hunk 8]
```diff
@@ -210,25 +237,17 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
...
-    return;
-  }
-
-  if (elem->is_aggregated() == true && in_flush_all_ == false) {
-    // Insert the element back to the cache while aggregator is waiting for the
-    // response.
-    elem->set_in_flight(true);
+  if (in_flush_all_ == false) {
+    // insert the element back to the cache
     cache_->Insert(elem->signature(), elem, 1);
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: not very nice to always keep this item in cache.   
How about comparing with last_refresh_time,  if it is too old,  such as 10 minutes, remove it from cache.

File: src/quota_aggregator_impl.h [hunk 9]
```diff
@@ -156,6 +179,8 @@ class QuotaAggregatorImpl : public QuotaAggregator,
   // Usually called at destructor.
   virtual ::google::protobuf::util::Status FlushAll();
 
+  bool ShouldRefresh(const CacheElem& elem);
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.h]: at "const" to the function.

File: include/aggregation_options.h [hunk 10]
```diff
@@ -31,15 +31,18 @@ typedef std::unordered_map<std::string,
...
+      expiration_interval_ms(600000){}
 
   // Constructor.
   // cache_entries is the maximum number of cache entries that can be kept in
   // the aggregation cache. Cache is disabled when cache_entries <= 0.
   // refresh_interval_ms is the maximum milliseconds before an aggregated quota
   // request needs to send to remote server again.
-  QuotaAggregationOptions(int cache_entries, int refresh_interval_ms)
-      : num_entries(cache_entries), refresh_interval_ms(refresh_interval_ms) {}
+  QuotaAggregationOptions(int cache_entries, int refresh_interval_ms,
```
Comment (qiwzhang) [inline include/aggregation_options.h]: Please add comment saying expiration_interval should be big.  It should be at lease 10 time bigger than rate limit window.

**Pull request merged**